Add unauthorised volume mode conversion blog article (#32419)
kubernetes/website: content/en/blog/_posts/2022-04-12-prevent-unauthorised-volume-mode-conversion.md

> ---

Reviewer: Date in the file name also needs to be updated.

Author: Moved the file too

layout: blog
title: 'Prevent unauthorised volume mode conversion'

> date: 2022-04-12

Reviewer 2: ```suggestion
date: 2022-05-20
```
(and update the path to match)

Reviewer: I believe it is 2022-05-18

Author: done!

backup operation, for efficiency purposes, which prevents Kubernetes from blocking
the operation completely and presents a challenge in distinguishing trusted
users from malicious ones.

### Preventing unauthorised users from converting the volume mode

In this context, an authorised user is one who has access rights to perform `Update` 
or `Patch` operations on `VolumeSnapshotContents`, which is a cluster-level resource.  
It is upto the cluster administrator to provide these rights only to trusted users
or applications, like backup vendors.

If the alpha feature is [enabled](https://kubernetes-csi.github.io/docs/) in 
`snapshot-controller` and `external-provisioner`, then unauthorised users will
not be allowed to modify the volume mode of a PVC when it is being created from
a `VolumeSnapshot`.

To convert the volume mode, an authorised user must do the following:

1. Identify the `VolumeSnapshot` that is to be used as the data source for a newly 
created PVC in the given namespace. 
2. Identify the `VolumeSnapshotContent` bound to the above `VolumeSnapshot`.

```yaml
$ kubectl get volumesnapshot -n <namespace>

> ```

Reviewer 2: This isn't YAML; also, please don't include a prompt.

Author: Fixed!

3. Add the annotation [`snapshot.storage.kubernetes.io/allowVolumeModeChange`](/content/en/docs/reference/labels-annotations-taints/)
to the `VolumeSnapshotContent`. 

4. This annotation can be added either via software or manually by the authorised

> user. The `VolumeSnapshotContent` must look like below after this change:

Reviewer 2: ```suggestion
   user. The `VolumeSnapshotContent` annotation must look like following manifest fragment:
```

Author: Changed

```yaml
kind: VolumeSnapshotContent 
metadata: 
  annotations: 
    - snapshot.storage.kubernetes.io/allowVolumeModeChange: "true"
...
```

> NOTE: For pre-provisioned `VolumeSnapshotContents`, the user has an additional 
> step of setting `spec.SourceVolumeMode` field to either `Filesystem` or `Block`,
> depending on the volume from which this snapshot was taken.

Reviewer 2: ```suggestion
**Note**: For pre-provisioned `VolumeSnapshotContents`, you must take an extra
step of setting `spec.sourceVolumeMode` field to either `Filesystem` or `Block`,
depending on the mode of the volume from which this snapshot was taken.
```

Author: Changed

An example is shown below:

```yaml
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshotContent
metadata:
  annotations:
  - snapshot.storage.kubernetes.io/allowVolumeModeChange: "true"
  name: new-snapshot-content-test
spec:
  deletionPolicy: Delete
  driver: hostpath.csi.k8s.io
  source:
    snapshotHandle: 7bdd0de3-aaeb-11e8-9aae-0242ac110002
  sourceVolumeMode: Filesystem
  volumeSnapshotRef:
    name: new-snapshot-test
    namespace: default
```

Repeat (1)-(3) for all `VolumeSnapshotContents` whose volume mode needs to be 
converted during a backup or restore operation.

If the above annotation is present on a `VolumeSnapshotContent` object, 
Kubernetes will not prevent the volume mode from being converted.
Users should keep this in mind before they attempt to add the annotation 
to any `VolumeSnapshotContent`. 

### What's next

[Enable this feature](https://kubernetes-csi.github.io/docs/) and let us know 
what you think!

We hope this feature causes no disruption to existing workflows while preventing
malicious users from exploiting security vulnerabilities in their clusters. 

For any issues, create a thread in the #sig-storage slack channel or an issue

> in the CSI external-snapshotter sidecar [repository](https://github.com/kubernetes-csi/external-snapshotter) 

Reviewer: Remove "sidecar".  This repo also has code for CRDs, snapshot controller, and snapshot validation webhook.

Author: makes sense, done!

> and assign it to [@RaunakShah](https://github.com/RaunakShah).

Reviewer: Can you remove the "assign it to ..." part?

Author: done!

---
layout: blog
title: 'Prevent unauthorised volume mode conversion'
date: 2022-04-12
slug: prevent-unauthorised-volume-mode-conversion-alpha
---

**Author:** Raunak Shah (Mirantis)

> Kubernetes v1.24 introduces a new alpha-level feature that prevents unauthorised users from modifying the volume mode of a [`PeristentVolumeClaim`](/docs/concepts/storage/persistent-volumes.md). This feature requires [`VolumeSnapshot`](/docs/concepts/storage/volume-snapshots.md) APIs with version `v6.0.0` onwards and `external-provisioner` version `v3.2.0` onwards.  

Reviewer 2: **I'm not sure what should go in place of** `FIXME` here.

```suggestion
Kubernetes v1.24 introduces a new alpha-level feature that prevents unauthorised users
from modifying the volume mode of a [PersistentVolumeClaim](/docs/concepts/storage/persistent-volumes/).
Using this new feature requires that you have added the
[VolumeSnapshot](/docs/concepts/storage/volume-snapshots/)
extension API into your cluster. You also need to be using FIXME version `v6.0.0` or later,
and `external-provisioner` version `v3.2.0` or later. 
```

Reviewer 2: Note that `PeristentVolumeClaim` is a typo.

### The problem

> As of Kubernetes 1.23, users can leverage the `VolumeSnapshot` feature, which GA'd in Kubernetes 1.20, to create a `PersistentVolumeClaim` (or `PVC`) from a previously taken `VolumeSnapshot`. This is done by pointing the `Spec.dataSource` parameter of the `PVC` to an existing `VolumeSnapshot` instance. 

Reviewer 2: ```suggestion
If you're using Kubernetes 1.23 or later, you can leverage the `VolumeSnapshot` API (stable since Kubernetes 1.20)
to create a new `PersistentVolumeClaim` (shortened as PVC) from a `VolumeSnapshot` that you had previously made.
You do this by defining the `spec.dataSource` parameter of the new PVC to refer to an existing
`VolumeSnapshot`. 
```

I wouldn't use code style for API kinds (PersistentVolumeClaim, etc). We usually don't in our docs.

> As of Kubernetes 1.23, users can leverage the `VolumeSnapshot` feature, which GA'd in Kubernetes 1.20, to create a `PersistentVolumeClaim` (or `PVC`) from a previously taken `VolumeSnapshot`. This is done by pointing the `Spec.dataSource` parameter of the `PVC` to an existing `VolumeSnapshot` instance. 

Reviewer: "Kubernetes 1.23" should be changed to "Kubernetes 1.20" as it is GA in 1.20.

Author: Fixed

There is no logic that validates whether the original volume mode of the `PVC`, whose snapshot was taken, matches the volume mode of the newly created `PVC`, that is being created from the existing `VolumeSnapshot`.

There is logic in allowing this, as many popular storage backup vendors convert the volume mode, during the course of a backup operation, for efficiency purposes.

> However this also presents a security gap that allows malicious users to potentially exploit an as-yet-unknown CVE in the kernel.

Reviewer 2: ```suggestion
However this also presents a security gap that allows malicious users to potentially exploit an as-yet-unknown flaw in a node's operating system.
```

Author: Changed to `vulnerability in the host operating system.`

### Preventing unauthorised users from converting the volume mode

> If the alpha feature is enabled in `snapshot-controller` and `external-provisioner`, then unauthorised users will not be allowed to modify the volume mode of a `PVC` when it is being created from a `VolumeSnapshot`.

Reviewer 2: ```suggestion
If you enable the alpha feature in both `snapshot-controller` and `external-provisioner`,
then unauthorised users will not be allowed to modify the volume mode of a `PVC` that
is being created from a `VolumeSnapshot`.
```
consider making “enable” a link to the docs page on how to do that

Author: I've added a link to external docs and have a PR out for review - https://github.com/kubernetes-csi/docs/pull/511

An unauthorised user is defined as one who does not have existing permissions to alter the cluster-scoped `VolumeSnapshotContent` resource.
Backup vendors normally have this permission on clusters where a backup is to be performed. 

To convert the volume mode, an authorised user must do the following:

1. Identify the `VolumeSnapshot` that is to be used as the data source for a newly created `PVC`. 
2. Identify the `VolumeSnapshotContent` bound to the above `VolumeSnapshot`.

> 3. Add a new annotation `snapshot.storage.kubernetes.io/allowVolumeModeChange` to the `VolumeSnapshotContent`. 

Reviewer 2: ```suggestion
3. Add the annotation `snapshot.storage.kubernetes.io/allowVolumeModeChange` to the `VolumeSnapshotContent`. 
```
No need to mention that what you're adding is new; that is implied.

Author: Done

> This annotation can be added either via software or manually by the backup vendor. The VolumeSnapshotContent must look like below after this change:

Reviewer 2: The backup tool vendor **manually** sets an annotation? Is that right?

Author: That is correct

```yaml
kind: VolumeSnapshotContent 
metadata: 
	annotations: 

> 		- snapshot.storage.kubernetes.io/allowVolumeModeChange: "true"

Reviewer 2: Are these tabs? I'd use spaces.

Author: Fixed

...
```

> NOTE: For pre-provisioned `VolumeSnapshotContents`, the user has an additional step of setting `Spec.SourceVolumeMode` field to either `Filesystem` or `Block`, depending on the volume from which this snapshot was taken.

Reviewer 2: ```suggestion
For pre-provisioned `VolumeSnapshotContents`, the user has an additional step of setting `spec.SourceVolumeMode` field to either `Filesystem` or `Block`, depending on the volume from which this snapshot was taken.
```

`Spec` is miscapitalized (check the manifest). I know the API definition uses PascalCase but the API itself uses camelCase.

Author: Done

An example is shown below:

```yaml
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshotContent
metadata:
  name: new-snapshot-content-test
spec:
  deletionPolicy: Delete
  driver: hostpath.csi.k8s.io
  source:
    snapshotHandle: 7bdd0de3-aaeb-11e8-9aae-0242ac110002
  sourceVolumeMode: Filesystem
  volumeSnapshotRef:
    name: new-snapshot-test
    namespace: default
```

Repeat (1)-(3) for all `VolumeSnapshotContent`s whose volume mode needs to be converted during a backup or restore operation.

If the above annotation is present on a `VolumeSnapshotContent` object, Kubernetes will not prevent the volume mode from being converted.
Users should keep this in mind before they attempt to add the annotation to any `VolumeSnapshotContent`. 

### How to enable the feature

This feature can be enabled by setting `prevent-volume-mode-conversion` flag to `true` in the `snapshot-controller` and `external-provisioner` spec, as shown below:

```yaml
apiVersion: apps/v1
kind: Deployment
...
spec:
      containers:
      - args:
        - --leader-election=true
        - --prevent-volume-mode-conversion=true
        image: snapshot-controller:v6.0.0
...

> ```

Reviewer 2: Consider moving this section to the docs and then hyperlinking there.

Author: I've added a link to external docs and have a PR out for review - https://github.com/kubernetes-csi/docs/pull/511
